File: backend/agent/tools/unlocker.py
"""
Phantom — Web Unlocker Tool (Glassdoor & News Signals)

Uses Bright Data Web Unlocker to access Glassdoor and Google News
for company-level sentiment signals.

Glassdoor approach: Search Google for the company's Glassdoor page
rather than trying to construct a direct Glassdoor URL (which requires
knowing the company's internal numeric ID). This is more reliable.

News approach: Google News search for company expansion/funding keywords.

Input: company name
Output: WebUnlockerSignals with glassdoor and news boolean flags + snippets
"""

import os
import re
import logging
import httpx
from urllib.parse import quote_plus
from typing import Optional
import asyncio

from models.schemas import WebUnlockerSignals

logger = logging.getLogger(__name__)

UNLOCKER_API_URL = "https://api.brightdata.com/request"

# Keywords indicating hiring freeze in Glassdoor reviews
FREEZE_KEYWORDS = [
    "hiring freeze", "freeze", "not hiring", "no new hires",
    "headcount freeze", "budget freeze", "paused hiring", "no headcount",
]
# Keywords indicating layoffs
LAYOFF_KEYWORDS = [
    "layoff", "laid off", "layoffs", "rif", "restructuring",
    "downsizing", "reduction in force", "let go", "workforce reduction",
    "job cuts", "mass layoff",
]
# Keywords indicating genuine expansion
EXPANSION_KEYWORDS = [
    "expansion", "expanding", "new office", "growing team",
    "scaling", "new market", "launch", "opened", "new location",
]
# Keywords indicating funding/investment
FUNDING_KEYWORDS = [
    "funding", "raised", "series a", "series b", "series c",
    "seed round", "investment", "ipo", "valuation", "venture",
    "capital raise", "million", "billion",
]


async def fetch_unlocker_signals(
    query: str,
    company: Optional[str] = None,
    **kwargs,
) -> Optional[WebUnlockerSignals]:
    """
    Fetch company sentiment signals from Glassdoor (via Google) and news.

    Makes two parallel requests:
    1. Google search for the company's Glassdoor reviews — checks freeze/layoff keywords
    2. Google News for the company — checks expansion/funding keywords

    Why Google search for Glassdoor instead of direct Glassdoor URL:
    - Glassdoor direct URLs require the company's internal numeric ID (e.g. E675258)
    - There is no reliable way to derive that ID from a company name
    - Googling "company site:glassdoor.com reviews" returns the correct page
      in the snippets, which is sufficient for keyword analysis

    Args:
        query: Search query context (used if company not specified)
        company: Company name to research

    Returns:
        WebUnlockerSignals with boolean flags and relevant snippets
    """
    api_key = os.getenv("BRIGHT_DATA_API_KEY")
    zone = os.getenv("BRIGHT_DATA_SERP_ZONE", "serp")  # Use SERP for Google search

    if not api_key:
        logger.error("BRIGHT_DATA_API_KEY not set")
        return None

    search_target = company or query

    # Run Glassdoor and news fetches in parallel
    glassdoor_content, news_content = await asyncio.gather(
        _fetch_glassdoor_via_google(search_target, api_key, zone),
        _fetch_news_via_google(search_target, api_key, zone),
    )

    signals = _analyse_signals(glassdoor_content, news_content)

    logger.info(
        f"Web Unlocker: freeze={signals.glassdoor_mentions_freeze}, "
        f"layoffs={signals.glassdoor_mentions_layoffs}, "
        f"expansion={signals.has_expansion_news}, "
        f"funding={signals.has_funding_news} "
        f"for: '{search_target}'"
    )

    return signals


async def _fetch_glassdoor_via_google(
    company: str,
    api_key: str,
    zone: str,
) -> Optional[str]:
    """
    Search Google for the company's Glassdoor reviews via SERP API.

    Returns Google search results page as markdown. The snippets shown
    by Google for Glassdoor results often contain review excerpts including
    phrases like "hiring freeze", "layoffs", "great company" etc.
    """
    encoded_query = quote_plus(f"{company} site:glassdoor.com reviews")
    search_url = f"https://www.google.com/search?q={encoded_query}&hl=en"

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }
    payload = {
        "zone": zone,
        "url": search_url,
        "format": "json",
        "brd_json": True,
    }

    try:
        async with httpx.AsyncClient(timeout=25.0) as client:
            response = await client.post(UNLOCKER_API_URL, json=payload, headers=headers)
            response.raise_for_status()
            try:
                data = response.json()
                # Extract snippets from organic results
                organic = data.get("organic", [])
                if isinstance(organic, list):
                    snippets = " ".join(
                        item.get("description", item.get("snippet", ""))
                        for item in organic[:5]
                    )
                    return snippets
                return str(data)
            except Exception:
                return response.text
    except Exception as e:
        logger.warning(f"Glassdoor Google search failed for '{company}': {e}")
        return None


async def _fetch_news_via_google(
    company: str,
    api_key: str,
    zone: str,
) -> Optional[str]:
    """
    Search Google News for recent company expansion or funding news via SERP API.
    """
    encoded_query = quote_plus(f"{company} hiring expansion funding 2024 2025")
    search_url = f"https://www.google.com/search?q={encoded_query}&tbm=nws&hl=en"

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {api_key}",
    }
    payload = {
        "zone": zone,
        "url": search_url,
        "format": "json",
        "brd_json": True,
    }

    try:
        async with httpx.AsyncClient(timeout=25.0) as client:
            response = await client.post(UNLOCKER_API_URL, json=payload, headers=headers)
            response.raise_for_status()
            try:
                data = response.json()
                organic = data.get("organic", data.get("news", []))
                if isinstance(organic, list):
                    snippets = " ".join(
                        item.get("description", item.get("snippet", ""))
                        for item in organic[:5]
                    )
                    return snippets
                return str(data)
            except Exception:
                return response.text
    except Exception as e:
        logger.warning(f"News Google search failed for '{company}': {e}")
        return None


def _analyse_signals(
    glassdoor_content: Optional[str],
    news_content: Optional[str],
) -> WebUnlockerSignals:
    """
    Keyword-scan the fetched content for ghost job indicators.

    Args:
        glassdoor_content: Text from Glassdoor Google search snippets
        news_content: Text from Google News search snippets

    Returns:
        WebUnlockerSignals with boolean flags and relevant snippets
    """
    mentions_freeze = False
    mentions_layoffs = False
    review_snippets: list[str] = []
    has_expansion_news = False
    has_funding_news = False
    news_snippets: list[str] = []

    if glassdoor_content:
        content_lower = glassdoor_content.lower()
        for kw in FREEZE_KEYWORDS:
            if kw in content_lower:
                mentions_freeze = True
                idx = content_lower.index(kw)
                snippet = glassdoor_content[max(0, idx - 80): idx + 100].strip()
                review_snippets.append(snippet)
                break
        for kw in LAYOFF_KEYWORDS:
            if kw in content_lower:
                mentions_layoffs = True
                idx = content_lower.index(kw)
                snippet = glassdoor_content[max(0, idx - 80): idx + 100].strip()
                review_snippets.append(snippet)
                break

    if news_content:
        content_lower = news_content.lower()
        for kw in EXPANSION_KEYWORDS:
            if kw in content_lower:
                has_expansion_news = True
                idx = content_lower.index(kw)
                snippet = news_content[max(0, idx - 80): idx + 100].strip()
                news_snippets.append(snippet)
                break
        for kw in FUNDING_KEYWORDS:
            if kw in content_lower:
                has_funding_news = True
                idx = content_lower.index(kw)
                snippet = news_content[max(0, idx - 80): idx + 100].strip()
                news_snippets.append(snippet)
                break

    return WebUnlockerSignals(
        glassdoor_mentions_freeze=mentions_freeze,
        glassdoor_mentions_layoffs=mentions_layoffs,
        glassdoor_review_snippets=review_snippets if review_snippets else None,
        recent_news=news_snippets if news_snippets else None,
        has_expansion_news=has_expansion_news,
        has_funding_news=has_funding_news,
        source="Web Unlocker",
    )

File: src/lib/agents/unlocker.ts
/**
 * Web Unlocker Agent — Fetches Glassdoor review signals and company news.
 *
 * Uses Bright Data SERP API to search Google for:
 * - Glassdoor reviews mentioning hiring freezes or layoffs
 * - Recent news about company expansion or funding
 *
 * These signals have Medium and Low weight in the ghost job scoring system.
 *
 * @module agents/unlocker
 */
import { getEnv } from "@/lib/config/env";
import {
  BRIGHT_DATA_API_URL,
  FREEZE_KEYWORDS,
  LAYOFF_KEYWORDS,
  EXPANSION_KEYWORDS,
  FUNDING_KEYWORDS,
} from "@/lib/data";
import { parseGoogleSerpHtml } from "@/lib/parsers/serp-html";
import type { WebUnlockerSignals } from "@/lib/types";

export async function fetchUnlockerSignals(query: string, company?: string): Promise<WebUnlockerSignals | null> {
  const env = getEnv();
  const searchTarget = company || query;
  const apiKey = env.BRIGHT_DATA_API_KEY;
  const zone = env.BRIGHT_DATA_SERP_ZONE;

  const [glassdoorContent, newsContent] = await Promise.all([
    fetchGlassdoorViaGoogle(searchTarget, apiKey, zone),
    fetchNewsViaGoogle(searchTarget, apiKey, zone),
  ]);

  const signals = analyseSignals(glassdoorContent, newsContent);

  console.log(
    `Web Unlocker: freeze=${signals.glassdoor_mentions_freeze}, layoffs=${signals.glassdoor_mentions_layoffs}, ` +
    `expansion=${signals.has_expansion_news}, funding=${signals.has_funding_news} for: "${searchTarget}"`
  );

  return signals;
}

async function fetchGlassdoorViaGoogle(company: string, apiKey: string, zone: string): Promise<string | null> {
  const encodedQuery = encodeURIComponent(`${company} site:glassdoor.com reviews`);
  const searchUrl = `https://www.google.com/search?q=${encodedQuery}&hl=en`;

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${apiKey}`,
  };

  const payload = { zone, url: searchUrl };

  try {
    const response = await fetch(BRIGHT_DATA_API_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(25_000),
    });

    if (!response.ok) return null;

    const html = await response.text();
    const parsed = parseGoogleSerpHtml(html, 5);
    return parsed.map((item) => `${item.title} ${item.snippet}`).join(" ");
  } catch (err) {
    console.warn(`Glassdoor Google search failed for "${company}": ${err}`);
    return null;
  }
}

async function fetchNewsViaGoogle(company: string, apiKey: string, zone: string): Promise<string | null> {
  // Dynamically compute year range so news results stay current
  const currentYear = new Date().getFullYear();
  const previousYear = currentYear - 1;
  const encodedQuery = encodeURIComponent(`${company} hiring expansion funding ${previousYear} ${currentYear}`);
  const searchUrl = `https://www.google.com/search?q=${encodedQuery}&tbm=nws&hl=en`;

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${apiKey}`,
  };

  const payload = { zone, url: searchUrl };

  try {
    const response = await fetch(BRIGHT_DATA_API_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(25_000),
    });

    if (!response.ok) return null;

    const html = await response.text();
    const parsed = parseGoogleSerpHtml(html, 5);
    return parsed.map((item) => `${item.title} ${item.snippet}`).join(" ");
  } catch (err) {
    console.warn(`News Google search failed for "${company}": ${err}`);
    return null;
  }
}

/** @internal Exported for testing */
export function analyseSignals(
  glassdoorContent: string | null,
  newsContent: string | null
): WebUnlockerSignals {
  let mentionsFreeze = false;
  let mentionsLayoffs = false;
  const reviewSnippets: string[] = [];
  let hasExpansionNews = false;
  let hasFundingNews = false;
  const newsSnippets: string[] = [];

  if (glassdoorContent) {
    const lower = glassdoorContent.toLowerCase();
    for (const kw of FREEZE_KEYWORDS) {
      if (lower.includes(kw)) {
        mentionsFreeze = true;
        const idx = lower.indexOf(kw);
        reviewSnippets.push(glassdoorContent.slice(Math.max(0, idx - 80), idx + 100).trim());
        break;
      }
    }
    for (const kw of LAYOFF_KEYWORDS) {
      if (lower.includes(kw)) {
        mentionsLayoffs = true;
        const idx = lower.indexOf(kw);
        reviewSnippets.push(glassdoorContent.slice(Math.max(0, idx - 80), idx + 100).trim());
        break;
      }
    }
  }

  if (newsContent) {
    const lower = newsContent.toLowerCase();
    for (const kw of EXPANSION_KEYWORDS) {
      if (lower.includes(kw)) {
        hasExpansionNews = true;
        const idx = lower.indexOf(kw);
        newsSnippets.push(newsContent.slice(Math.max(0, idx - 80), idx + 100).trim());
        break;
      }
    }
    for (const kw of FUNDING_KEYWORDS) {
      if (lower.includes(kw)) {
        hasFundingNews = true;
        const idx = lower.indexOf(kw);
        newsSnippets.push(newsContent.slice(Math.max(0, idx - 80), idx + 100).trim());
        break;
      }
    }
  }

  return {
    glassdoor_mentions_freeze: mentionsFreeze,
    glassdoor_mentions_layoffs: mentionsLayoffs,
    glassdoor_review_snippets: reviewSnippets.length > 0 ? reviewSnippets : null,
    recent_news: newsSnippets.length > 0 ? newsSnippets : null,
    has_expansion_news: hasExpansionNews,
    has_funding_news: hasFundingNews,
    source: "Web Unlocker",
  };
}
